checker/env.go

```diff
@@ -57,19 +58,48 @@ func (e *Env) Add(decls ...*checked.Decl) {
 	}
 }
 
-func (e *Env) addFunction(decl *checked.Decl) {
-	current := e.declarations.FindFunction(decl.Name)
-	if current != nil {
-		if current.Name != decl.Name {
+func (e *Env) addOverload(f *checked.Decl, overload *checked.Decl_FunctionDecl_Overload) {
+	function := f.GetFunction()
+	emptyMappings := types.NewMapping()
+	overloadFunction := types.NewFunction(overload.GetResultType(),
+		overload.GetParams()...)
+	overloadErased := types.Substitute(emptyMappings, overloadFunction, true)
+	for _, existing := range function.GetOverloads() {
+		existingFunction := types.NewFunction(existing.GetResultType(),
+			existing.GetParams()...)
+		existingErased := types.Substitute(emptyMappings, existingFunction, true)
+		overlap := (types.IsAssignable(emptyMappings, overloadErased, existingErased) != nil ||
+			types.IsAssignable(emptyMappings, existingErased, overloadErased) != nil)
+		if overlap &&
+			overload.GetIsInstanceFunction() == existing.GetIsInstanceFunction() {
+			e.errors.overlappingOverload(common.NoLocation, f.Name, overloadFunction,
+				existingFunction)
+			return
+		}
+	}
+
+	for _, macro := range parser.AllMacros {
+		if macro.GetName() == f.Name && macro.GetIsInstanceStyle() == overload.GetIsInstanceFunction() &&
+			macro.GetArgCount() == len(overload.GetParams()) {
+			e.errors.overlappingMacro(common.NoLocation, f.Name, macro.GetArgCount())
 			return
 		}
-		// TODO: Check for conflicts.
-		function := current.GetFunction()
-		function.Overloads = append(function.Overloads,
-			decl.GetFunction().Overloads...)
-		decl = current
-	} else {
-		e.declarations.AddFunction(decl)
+	}
+	function.Overloads = append(function.GetOverloads(), overload)
+}
+
+func (e *Env) addFunction(decl *checked.Decl) {
+	current := e.declarations.FindFunction(decl.Name)
+	if current == nil {
+		//Add the function declaration without overloads and check the overloads below.
+		current = decls.NewFunction(decl.Name)
+		e.declarations.AddFunction(current)
+	}
+	if current.Name != decl.Name {
```

Comment:
Under what circumstance would the current.Name and decl.Name not be equal? Consider removing this if-block

Reply:
I mostly had left this here because it was in the original changes. I agree it looks useless. Removed it.